### DISCUSSION TITLE
How to mock import.meta.glob()?

### DISCUSSION THREAD

**Discussion by vekunz** [vitest-dev/vitest#3564] [Q&A] [answered]
Hello, I have a very special mocking question.

We have code, that dynamically imports files with a glob pattern. We use this statement for this:

```ts
const modules = import.meta.glob('somewhere/*/*.js');
```

Now I have to write unit tests for this code. I tried several ways to mock the `import.meta.glob()` call but without success. I tried these options:

- `vi.mock('somewhere/folder/file.js', () => ...)`
- `vi.spyOn(import.meta, 'glob').mockReturnValue(...)`
- `import.meta.glob = vi.fn().mockReturnValue(...)`
- `vi.stubGlobal('import', {meta: {glob: vi.fn().mockReturnValue(...)}})`

The last three I executed in a `vi.hoisted()` statement.

In any case, the mock did not work. While executing the test, `import.meta.glob()` imported the actual files from the file system.

Is there any way to directly mock `import.meta.glob()`?

**Comment by eloiqs** [comment]
We also aren't able to mock `import.meta.env.DEV` value like this
```
Object.defineProperty(import.meta.env, 'DEV', { value: false });
```
This works though:
```
Object.defineProperty(import.meta.env, 'NODE_ENV', { value: 'production' });
```

**Reply by sheremet-va** [reply]
You can just assign it:

```ts
import.meta.env.NODE_ENV = 'production'
```

**Comment by luiguild** [comment]
Just include in your `tsconfig.json`
```
"compilerOptions": {
    "types": [
      //... your types
      "vitest/globals"
    ],
    "module": "ES2022" // can be other versions too
}
```

After that you supposed to be able to use the `import.meta.glob` in your tests.

**Reply by vekunz** [reply]
Do you mean `vitest/globals`? We already have this. But we don't want to use `import.meta.glob` but we want to mock it.

**Comment by sheremet-va** [comment, marked as answer]
You cannot mock it, `import.meta.glob` doesn't exist at runtime, it's a compiler hint. You need to mock all modules that it imports one by one.

**Reply by ricardovanlaarhoven** [reply]
But is there a way to test a function that depends on it?

**Comment by AnshulSurana** [comment]
I also want to mock in vitest, import.meta.client and import.meta.server, these two seems to be read-only values from Nuxt 3.
But i was able to mock process.client and process.server.

**Comment by simonesada** [comment]
Has anyone found a solution?

**Reply by binhtran04** [reply]
Firstly, move the `import.meta.glob` into a separate module. For example 
```
// lab-loader.ts
export const labModules = import.meta.glob(
  "/src/assets/lab-instructions/**/*.md",
  {
    query: "?raw",
    import: "default",
  }
);
```
Then, in the test setup, mock the whole module
```
// Use vi.hoisted to ensure mock storage is available during module evaluation
const mockGlob = vi.hoisted(() => ({
  modules: {} as Record<string, () => Promise<string>>,
}));

// Mock the lab-loader module to use our test data
vi.mock("./lab-loader", () => ({
  labModules: mockGlob.modules,
}));
```

Finally, in the tests

```
const key = `/src/assets/lab-instructions/${labId}.md`;
mockGlob.modules[key] = () => Promise.resolve(content);
```

Remember to reset the mock on before each
```
Object.keys(mockGlob.modules).forEach((key) => delete mockGlob.modules[key]);

```
